packages/terra-dropdown-button/src/DropdownButton.jsx

```diff
@@ -57,29 +57,62 @@ class DropdownButton extends React.Component {
     this.handleDropdownRequestClose = this.handleDropdownRequestClose.bind(this);
     this.handleKeyDown = this.handleKeyDown.bind(this);
     this.handleKeyUp = this.handleKeyUp.bind(this);
+    this.setButtonNode = this.setButtonNode.bind(this);
+    this.getButtonNode = this.getButtonNode.bind(this);
+    this.setListNode = this.setListNode.bind(this);
+    this.state = { isOpen: false, isActive: false, openedViaKeyboard: false };
+  }
+
+  setListNode(element) {
+    this.dropdownList = element;
+  }
 
-    this.state = { isOpen: false, isActive: false };
+  setButtonNode(node) {
+    this.buttonNode = node;
   }
 
-  handleDropdownButtonClick() {
+  getButtonNode() {
+    return this.buttonNode;
+  }
+
+  handleDropdownButtonClick(event) {
+    if (this.state.isOpen) {
+      this.setState({ openedViaKeyboard: false });
+    }
     this.setState(prevState => ({ isOpen: !prevState.isOpen }));
+    // See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Button#Clicking_and_focus
+    // Button on Firefox, Safari and IE running on OS X does not receive focus when clicked.
+    // This will put focus on the button when clicked.
+    event.currentTarget.focus();
   }
 
   handleDropdownRequestClose(callback) {
-    this.setState({ isOpen: false }, typeof callback === 'function' ? callback : undefined);
+    const onSelectCallback = typeof callback === 'function' ? callback : undefined;
+    this.setState({ isOpen: false, openedViaKeyboard: false, isActive: false }, onSelectCallback);
   }
 
-  /*
-    In FireFox active styles don't get applied on space
-   */
   handleKeyDown(event) {
-    if (event.keyCode === KeyCode.KEY_SPACE) {
-      this.setState({ isActive: true });
+    if (event.keyCode === KeyCode.KEY_SPACE || event.keyCode === KeyCode.KEY_RETURN) {
+      // In FireFox active styles don't get applied on space
+      this.setState({ isActive: true, openedViaKeyboard: true });
```

Comment:
I'm seeing strange keyboard interactions when using the space and enter keys.

The dropdown open and closes when opened by long pressing the enter key.

1. Open the dropdown-button example page and use tab to navigate and focus a dropdown button.
2. Press and hold enter.
3. Dropdown opens
4. Release enter.
5. Observe the dropdown closes.

Repeat the steps above using spacebar instead of enter and the results will be different. Some of this is likely native browser behavior with keyboard interactions but the set state calls happening seem to put the component into a bad state.

Reply:
It looked like default behaviour for RETURN Key press in html button that triggers click event of button. but when RETURN key was held for longer time keydown event and onClick event was being executed multiple times. this was causing dropdown to open and close by itself when  RETURN key was held for longer time.

Incase of Space key press keyup event was called following onClick callback which made dropdown to open after removing hold from space key. thus causing dropdown to open as expected.

hence prevented click call back in keyDown event and moved open logic to key-up event.